### PR TITLE
Fix issue#47: update exam overview after changing tabs

### LINKED ISSUE #47: Question's BestRecord is not updated when go back to the exam overview page
## Reproduce:

Flow:  `exams/1` -> `exams/1/problems/3/submissions`  -> upload a code -> back to   `exams/1`  
then you won't see the question's status updated

### PR DESCRIPTION
**Pull request by edisonhello**
Resolves #47.

### CODE REVIEW

File: src/app/exam/exam-questions/exam-questions.component.ts [hunk 1]
```diff
@@ -32,18 +32,22 @@ export class ExamQuestionsComponent implements OnInit, AfterViewInit {
...
+    this.refetchExam();
+  }
+
+  private refetchExam() {
+    const exam$ = this.http.get<ExamOverview>(`${this.baseUrl}/api/exams/${this.examId}/students/${this.studentService.currentStudent.id}/overview`);
+
     setTimeout(() => {
-      this.exam$.subscribe(e => {
+      exam$.subscribe(e => {
+        console.log('subscribe', e);
```
Comment (Johnny850807) [inline src/app/exam/exam-questions/exam-questions.component.ts:47]: remove comment

File: src/app/exam/exam-questions/exam-questions.component.ts [hunk 2]
```diff
@@ -14,14 +13,15 @@ import {initHighlight, parseMarkdown} from 'src/utils/markdownUtils';
 })
 export class ExamQuestionsComponent implements OnInit, AfterViewInit {
 
-  constructor(private examService: ExamService,
-              private studentService: StudentService,
-              private route: ActivatedRoute,
-              private router: Router,
-              private renderer: Renderer2) {
+  constructor(private http: HttpClient,
```
Comment (Johnny850807) [inline src/app/exam/exam-questions/exam-questions.component.ts:16]: don't perform any HTTP requests in the component, do it in the service.

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

File: src/app/exam/exam-questions/exam-questions.component.ts (hunk 2 above)
Comment (Johnny850807) [reply src/app/exam/exam-questions/exam-questions.component.ts:16]: @edisonhello this is not resolved?

should remove the dependency on `HttpClient`

File: src/app/services/Services.ts [hunk 3]
```diff
@@ -124,7 +124,7 @@ export abstract class ProblemService {
 })
 export abstract class ExamService {
 
-  abstract getExamProgressOverview(studentId: number, examId: number): Observable<ExamOverview>;
+  abstract getExamProgressOverview(studentId: number, examId: number, cached?: boolean): Observable<ExamOverview>;
```
Comment (Johnny850807) [inline src/app/services/Services.ts:127]: I think we don't need to cache it here, so just leave the non-cache implementation.

File: src/app/services/impl/HttpExamService.ts [hunk 4]
```diff
@@ -20,8 +20,12 @@ export class HttpExamService extends ExamService {
...
   }
 
-  getExamProgressOverview(studentId: number, examId: number): Observable<ExamOverview> {
-    return this.httpRequestCache.get(`${this.baseUrl}/api/exams/${examId}/students/${studentId}/overview`);
+  getExamProgressOverview(studentId: number, examId: number, cached: boolean = true): Observable<ExamOverview> {
+    if (cached) {
+      return this.httpRequestCache.get(`${this.baseUrl}/api/exams/${examId}/students/${studentId}/overview`);
+    } else {
+      return this.http.get<ExamOverview>(`${this.baseUrl}/api/exams/${examId}/students/${studentId}/overview`);
+    }
```
Comment (Johnny850807) [inline src/app/services/impl/HttpExamService.ts:28]: ditto, we don't cache for sure.

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done.

File: src/app/exam/exam-questions/exam-questions.component.ts (hunk 2 above)
Comment (edisonhello) [reply src/app/exam/exam-questions/exam-questions.component.ts]: Bruh, I forget to push.

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Pull request merged**